### PR TITLE
fix: use a more specific regex for jinja2 set/for statements

### PR DESCRIPTION
**Pull request by beckermr**
### Description

This PR uses a more specific regex for jinja2 set statements per the comment from @isuruf. 

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](https://github.com/conda/conda-build/blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

### CODE REVIEW

**Comment by beckermr** [comment]
pre-commit.ci autofix

File: conda_build/variants.py [hunk 1]
```diff
@@ -760,7 +760,11 @@ def find_used_variables_in_text(variant, recipe_text, selectors_only=False):
         # TODO: this `for` regex won't catch some common cases like lists of vars, multiline
         # jinja2 blocks, if filters on the for loop, etc.
         for_regex = r"(?:^|[^\{])\{%\s*for\s*.*\s*in\s*" + v_regex + r"(?:[^%]*?)?%\}"
```
Comment (isuruf) [inline conda_build/variants.py:762]: You'll need to fix this too

Comment (beckermr) [inline conda_build/variants.py:762]:
```suggestion
        for_regex = r"(?:^|[^\{])\{%\s*for\s*.*\s*in\s*" + v_regex + r"(?![a-zA-Z_0-9])(?:[^%]*?)?%\}"
```

File: news/5514-better-jinja2-regex.rst [hunk 2]
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* <news item>
+
+### Bug fixes
+
+* Fixed regex for jinja2 set statements to be more specific. (#5514)
```
Comment (beckermr) [inline news/5514-better-jinja2-regex.rst:7]:
```suggestion
* Fixed regex for jinja2 set/for statements to be more specific. (#5514)
```

**Review by isuruf — APPROVED** [review]

**Comment by beckermr** [comment]
Thanks for noticing and suggesting a fix @isuruf!

File: tests/requirements-ci.txt [hunk 3]
```diff
@@ -14,6 +14,6 @@ pytest-forked
 pytest-mock
 pytest-rerunfailures
 pytest-xdist
-ruamel.yaml
+ruamel.yaml <0.18.0  # temporary for older setuptools
```
Comment (beckermr) [inline tests/requirements-ci.txt:17]:
```suggestion
ruamel.yaml
```

**Comment by beckermr** [comment]
blocked by https://github.com/conda/conda-build/issues/5515

**Comment by beckermr** [comment]
I think this will cause bugs with statements like `{% set blah = var.upper() %}`.

**Comment by wolfv** [comment]
Can you not use real jinja functions for this?

**Comment by beckermr** [comment]
See the comments in the code. We had this discussion.

**Comment by beckermr** [comment]
Nvm. This works.

**Comment by beckermr** [comment]
@isuruf @conda/builds-tools this is ready to merge!

File: news/5514-better-jinja2-regex.rst [hunk 4]
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* <news item>
+
+### Bug fixes
+
+* Fixed regex for jinja2 set/for statements to be more specific. (#5514)
```
Comment (jaimergp) [inline news/5514-better-jinja2-regex.rst:7]:
```suggestion
* Fixed regex for Jinja2 `set` / `for` statements to be more specific. (#5514)
```

**Review by jaimergp — APPROVED** [review]

**Pull request merged**